README.md
_The Berkeley Tree DataBase (BTrDB) is pronounced "**Better DB**"._

# BTrDB

__A next-gen timeseries database for high-precision, high-sample-rate telemetry.__

__Problem:__ Existing timeseries databases are poorly equipped for a new
generation of ultra-fast sensor telemetry. Specifically, millions of
high-precision power meters are to be deployed throughout the power grid to help
analyze and prevent blackouts. Thus, new software must be built to facilitate
the storage and analysis of its data.

__Baseline:__ We need 1.4M inserts/s and 5x that in reads if we are to support
1000 [micro-synchrophasors] per server node.  No timeseries database can do
this.

[micro-synchrophasors]:https://arxiv.org/abs/1605.02813

## Summary

__Goals:__ Develop a multi-resolution storage and query engine for many 100+ Hz
streams at nanosecond precision—and operate at the full line rate of
underlying network or storage infrastructure for affordable cluster sizes (less
than six).

Developed at Berkeley, BTrDB offers new ways to support the aforementioned high
throughput demands and allows efficient querying over large ranges.

**Fast writes/reads**

Measured on a 4-node cluster (large EC2 nodes):

- 53 million inserted values per second
- 119 million queried values per second

**Fast analysis**

In under _200ms_, it can query a year of data at nanosecond-precision (2.1
trillion points) at any desired window—returning statistical summary points at any
desired resolution (containing a min/max/mean per point).

![zoom](https://user-images.githubusercontent.com/116838/34450616-5090c8a2-ecd3-11e7-8722-f5d7f131e909.gif)

**High compression**

Data is compressed by 2.93x—a significant improvement for high-precision
nanosecond streams. To achieve this, a modified version of _run-length encoding_
was created to encode the _jitter_ of delta values rather than the delta values
themselves.  Incidentally, this  outperforms the popular audio codec [FLAC]
which was the original inspiration for this technique.

[FLAC]:https://xiph.org/flac/

**Efficient Versioning**

Data is version-annotated to allow queries of data as it existed at a certain
time.  This allows reproducible query results that might otherwise change due
to newer realtime data coming in.  Structural sharing of data between versions
is done to make this process as efficient as possible.

## The Tree Structure

BTrDB stores its data in a time-partitioned tree.

All nodes represent a given time slot. A node can describe all points within
its time slot at a resolution corresponding to its depth in the tree.

The root node covers ~146 years. With a branching factor of 64, bottom nodes at
ten levels down cover 4ns each.

| level | node width                       |
|:------|:---------------------------------|
| 1     | 2<sup>62</sup> ns  (~146 years)  |
| 2     | 2<sup>56</sup> ns  (~2.28 years) |
| 3     | 2<sup>50</sup> ns  (~13.03 days) |
| 4     | 2<sup>44</sup> ns  (~4.88 hours) |
| 5     | 2<sup>38</sup> ns  (~4.58 min)   |
| 6     | 2<sup>32</sup> ns  (~4.29 s)     |
| 7     | 2<sup>26</sup> ns  (~67.11 ms)   |
| 8     | 2<sup>20</sup> ns  (~1.05 ms)    |
| 9     | 2<sup>14</sup> ns  (~16.38 µs)   |
| 10    | 2<sup>8</sup> ns   (256 ns)      |
| 11    | 2<sup>2</sup> ns   (4 ns)        |

A node starts as a __vector node__, storing raw points in a vector of size 1024.
This is considered a leaf node, since it does not point to any child nodes.

```
┌─────────────────────────────────────────────────────────────────┐
│                                                                 │
│                           VECTOR NODE                           │
│                     (holds 1024 raw points)                     │
│                                                                 │
├─────────────────────────────────────────────────────────────────┤
│ . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . │ <- raw points
└─────────────────────────────────────────────────────────────────┘
```

Once this vector is full and more points need to be inserted into its time slot,
the node is converted to a __core node__ by time-partitioning itself into 64
"statistical" points.

```
┌─────────────────────────────────────────────────────────────────┐
│                                                                 │
│                            CORE NODE                            │
│                   (holds 64 statistical points)                 │
│                                                                 │
├─────────────────────────────────────────────────────────────────┤
│ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ ○ │ <- stat points
└─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┘
  ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼ ▼  <- child node pointers
```

A __statistical point__ represents a 1/64 slice of its parent's time slot. It
holds the min/max/mean/count of all points inside its time slot, and points to a
new node holding extra details.  When a vector node is first converted to a core
node, the raw points are pushed into new vector nodes pointed to by the new
statistical points.

| level | node width                       | stat points per node | stat point width                 |
|:------|:---------------------------------|:---------------------|:---------------------------------|
| 1     | 2<sup>62</sup> ns  (~146 years)  | 2<sup>6</sup> (64)   | 2<sup>56</sup> ns  (~2.28 years) |
| 2     | 2<sup>56</sup> ns  (~2.28 years) | 2<sup>6</sup> (64)   | 2<sup>50</sup> ns  (~13.03 days) |
| 3     | 2<sup>50</sup> ns  (~13.03 days) | 2<sup>6</sup> (64)   | 2<sup>44</sup> ns  (~4.88 hours) |
| 4     | 2<sup>44</sup> ns  (~4.88 hours) | 2<sup>6</sup> (64)   | 2<sup>38</sup> ns  (~4.58 min)   |
| 5     | 2<sup>38</sup> ns  (~4.58 min)   | 2<sup>6</sup> (64)   | 2<sup>32</sup> ns  (~4.29 s)     |
| 6     | 2<sup>32</sup> ns  (~4.29 s)     | 2<sup>6</sup> (64)   | 2<sup>26</sup> ns  (~67.11 ms)   |
| 7     | 2<sup>26</sup> ns  (~67.11 ms)   | 2<sup>6</sup> (64)   | 2<sup>20</sup> ns  (~1.05 ms)    |
| 8     | 2<sup>20</sup> ns  (~1.05 ms)    | 2<sup>6</sup> (64)   | 2<sup>14</sup> ns  (~16.38 µs)   |
| 9     | 2<sup>14</sup> ns  (~16.38 µs)   | 2<sup>6</sup> (64)   | 2<sup>8</sup> ns   (256 ns)      |
| 10    | 2<sup>8</sup> ns   (256 ns)      | 2<sup>6</sup> (64)   | 2<sup>2</sup> ns   (4 ns)        |
| 11    | 2<sup>2</sup> ns   (4 ns)        | (no stat points at bottom) |  |

The sampling rate of the data at different moments will determine how deep the
tree will be during those slices of time. Regardless of the depth of the actual
data, the time spent querying at some higher level (lower resolution) will
remain fixed (quick) due to summaries provided by parent nodes.

...

## Appendix

This page is written based on the following sources:

- [Homepage](http://btrdb.io/)
- [Whitepaper](https://www.usenix.org/system/files/conference/fast16/fast16-papers-andersen.pdf)
- [Code](https://github.com/BTrDB/btrdb-server)


CHANGES.md
## v2018.04.30

[demo permalink][demo-2018-04-30]

* Fix nanosecond time scale ticks

[demo-2018-04-30]: http://btrdb-viz-2018-04-30.surge.sh

## v2018.03.19

[demo permalink][demo-2018-03-19]

* Replace line plots with bounding box plots, to clarify mid-res cell relationships (See [#12](https://github.com/PingThingsIO/btrdb-explained/pull/12))

[demo-2018-03-19]: http://btrdb-viz-2018-03-19.surge.sh

## v2018.02.08

[screenshot][screenshot-2018-02-08] | [demo permalink][demo-2018-02-08]

* Populate tree with Simplex Noise and plot (See [#7](https://github.com/PingThingsIO/btrdb-explained/pull/7))

[screenshot-2018-02-08]: https://user-images.githubusercontent.com/116838/35985837-2b30fb54-0cbd-11e8-86f7-705ced580456.gif
[demo-2018-02-08]: http://btrdb-viz-2018-02-08.surge.sh

## v2018.01.28

[screenshot][screenshot-2018-01-28] | [demo permalink][demo-2018-01-28]

* Show mid-resolution cells (See [#5](https://github.com/PingThingsIO/btrdb-explained/pull/5))

[screenshot-2018-01-28]: https://user-images.githubusercontent.com/116838/35482932-55640df4-0401-11e8-932a-da14fa03b7ae.gif
[demo-2018-01-28]: http://btrdb-viz-2018-01-28.surge.sh

## v2018.01.22

[screenshot][screenshot-2018-01-22] | [demo permalink][demo-2018-01-22]

* Add mouse controls for exploring the tree (See [#4](https://github.com/PingThingsIO/btrdb-explained/pull/4))

[screenshot-2018-01-22]: https://user-images.githubusercontent.com/116838/35245661-a2e4ad26-ff89-11e7-83a2-4db239a4ed4a.gif
[demo-2018-01-22]: http://btrdb-viz-2018-01-22.surge.sh

## v2018.01.11

[screenshot][screenshot-2018-01-11] | [demo permalink][demo-2018-01-11]

* Add calendar metaphor for a zoomable 2D representation (See [#3](https://github.com/PingThingsIO/btrdb-explained/pull/3))

[screenshot-2018-01-11]: https://user-images.githubusercontent.com/116838/34836478-9ed6755e-f6bd-11e7-8895-353dfdfbc2cc.gif
[demo-2018-01-11]: http://btrdb-viz-2018-01-11.surge.sh

## v2018.01.08

[screenshot][screenshot-2018-01-08] | [demo permalink][demo-2018-01-08]

* Add date and time annotations (See [#2](https://github.com/PingThingsIO/btrdb-explained/pull/2))

[screenshot-2018-01-08]: https://user-images.githubusercontent.com/116838/34710929-d4f974e2-f4e2-11e7-8ccf-87fda093b2a5.gif
[demo-2018-01-08]: http://btrdb-viz-2018-01-08.surge.sh/

## v2018.01.07

[screenshot][screenshot-2018-01-07] | [demo permalink][demo-2018-01-07]

* Mouse up and down to show zooming into random nodes of the tree (See [#1](https://github.com/PingThingsIO/btrdb-explained/pull/1))

[screenshot-2018-01-07]: https://user-images.githubusercontent.com/116838/34665780-2b68bb38-f427-11e7-96c1-95c80ed3f39c.gif
[demo-2018-01-07]: http://btrdb-viz-2018-01-07.surge.sh/


src/Viz.js
import React, { Component } from "react";
import * as d3scale from "d3-scale";
import { scaleTimeNano } from "./scaleTimeNano";
import * as d3interpolate from "d3-interpolate";
import * as d3ease from "d3-ease";
import * as d3transition from "d3-transition";
import hexToRgba from "hex-to-rgba";
// import * as d3shape from "d3-shape";

const nodeLengthLabels = [
  "146 years",
  "2.28 years",
  "13.03 days",
  "4.88 hours",
  "4.58 min",
  "4.29 s",
  "67.11 ms",
  "1.05 ms",
  "16.38 µs",
  "256 ns",
  "4 ns"
];
const theme = {
  green: "#1eb7aa",
  orange: "#db7b35"
};
const colors = {
  cellFillExpanded: "rgba(80,100,120, 0.15)",
  cellFillHighlight: hexToRgba(theme.green, 0.4),
  cellWall: hexToRgba("#555", 0.1),
  cellWallExpanded: "#555",
  cellWallHighlight: theme.green,

  unixEpoch: theme.green,
  now: theme.orange,
  dateTick: "rgba(90,110,100, 0.5)",
  scrub: "#e7e8e9",
  zoomCone: "rgba(80,100,120, 0.15)"
};

class Viz extends Component {
  constructor(props) {
    super(props);
    this.state = {
      // canvas
      width: 1024,
      height: 720,

      numCells: 64, // cells in a tree row
      numSquareCells: 8, // cells in a calendar row

      path: [0], // first of path is ignored for generality
      pathAnim: 1, // float representing what index of the path we are showing

      // Tree placement and sizing
      treeCellW: 8,
      treeCellH: 12,
      treeX: 40,
      treeY: 34,
      levelOffset: 5,
      cellHighlight: {},

      // Calendar placement and sizing
      calCellSize: 38,
      calX: 700,
      calY: 40,

      rootStart: -1152921504606846976,
      rootResolution: 56
    };
    for (let i = 1; i < 10; i++) {
      this.state.path.push(Math.floor(Math.random() * this.state.numCells));
    }
    this.state.pathAnim = this.state.path.length;
    this.createD3Objects();
  }
  componentWillMount() {
    this.computeDerivedState(this.props, this.state);
  }
  componentWillUpdate(nextProps, nextState) {
    this.computeDerivedState(nextProps, nextState);
  }
  componentDidMount() {
    window.addEventListener("keydown", this.onKeyDown);
    window.addEventListener("keyup", this.onKeyUp);
  }
  componentWillUnmount() {
    window.removeEventListener("keydown", this.onKeyDown);
    window.removeEventListener("keyup", this.onKeyUp);
  }
  createD3Objects = () => {
    this.d3 = {};
  };
  computeDerivedState = (props, state) => {
    const {
      treeCellW,
      treeCellH,
      calCellSize,
      path,
      rootStart,
      rootResolution,
      numCells,
      numSquareCells,
      levelOffset
    } = state;

    // `dip` is how long `t` will spend dropping the node before expanding it.
    const dipTime = 1 / levelOffset;

    const pixelRatio = window.devicePixelRatio || 1;

    // maps tree cell units to pixels
    const treeColX = d3scale
      .scaleLinear()
      .domain([0, 1])
      .range([0, treeCellW]);
    const treeRowY = d3scale
      .scaleLinear()
      .domain([0, 1])
      .range([0, treeCellH]);

    // maps calendar cell units to pixels
    const calColX = d3scale
      .scaleLinear()
      .domain([0, 1])
      .range([0, calCellSize]);
    const calRowY = d3scale
      .scaleLinear()
      .domain([0, 1])
      .range([0, calCellSize]);

    const treeW = treeCellW * numCells;
    const calW = calCellSize * numSquareCells;

    // compute scale for each level
    const pw = res => Math.pow(2, res);
    let start = rootStart;
    let res = rootResolution;
    const treeTimeX = [];
    const calTimeK = [];
    for (let i = 0; i < path.length; i++) {
      const end = start + pw(res) * numCells;
      const domain = [start, end];
      treeTimeX.push(
        scaleTimeNano()
          .domain(domain)
          .range([0, treeW])
      );
      calTimeK.push(
        scaleTimeNano()
          .domain(domain)
          .range([0, calW * numSquareCells])
      );
      if (i + 1 < path.length) {
        start += pw(res) * path[i + 1];
      }
      res -= 6;
    }

    // calendar time to x and row
    const calKX = k => k % calW;
    const calKRow = k => Math.floor(k / calW);

    this.ds = {
      pixelRatio,
      treeW,
      calW,
      treeColX,
      treeRowY,
      treeTimeX,
      calColX,
      calRowY,
      calTimeK,
      calKX,
      calKRow,
      dipTime
    };
  };
  drawTreeCell = (ctx, level, cell) => {
    const { treeCellH, treeCellW, cellHighlight } = this.state;
    const expanded = this.isCellExpanded(level, cell);
    const highlighted =
      cellHighlight &&
      cellHighlight.cell === cell &&
      cellHighlight.level === level;
    if (highlighted) {
      ctx.strokeStyle = colors.cellWallHighlight;
      ctx.fillStyle = colors.cellFillHighlight;
    } else if (expanded) {
      ctx.strokeStyle = colors.cellWallExpanded;
      ctx.fillStyle = colors.cellFillExpanded;
    } else {
      ctx.strokeStyle = colors.cellWall;
      ctx.fillStyle = "rgba(0,0,0,0)";
    }
    ctx.fillRect(0, 0, treeCellW, treeCellH);
    ctx.strokeRect(0, 0, treeCellW, treeCellH);
  };
  drawTreeNode = (ctx, level) => {
    const {
      numCells,
      treeCellW,
      treeCellH,
      path,
      levelOffset,
      pathAnim
    } = this.state;
    const { treeColX, treeRowY, treeTimeX, dipTime } = this.ds;

    // index of the parent
    // (where in the previous level our node is coming from)
    const parent = path[level];

    // `t` controls the animation of this node
    const t = d3scale
      .scaleLinear()
      .domain([level, level + 1])
      .range([0, 1])
      .clamp(true)(pathAnim);

    // do not draw if pathAnim has not reached our level
    if (t === 0) return;

    const domain = [0, dipTime, 1];

    // left side
    const scaleX0 = d3scale
      .scaleLinear()
      .domain(domain)
      .range([treeColX(parent), treeColX(parent), treeColX(0)])
      .clamp(true);

    // right side
    const scaleX1 = d3scale
      .scaleLinear()
      .domain(domain)
      .range([treeColX(parent + 1), treeColX(parent + 1), treeColX(numCells)])
      .clamp(true);

    // top side
    const scaleY = d3scale
      .scaleLinear()
      .domain(domain)
      .range([
        treeRowY(-levelOffset),
        treeRowY(-levelOffset * (1 - dipTime)),
        treeRowY(0)
      ])
      .clamp(true);

    // compute
    const x0 = scaleX0(t);
    const x1 = scaleX1(t);
    const y = scaleY(t);
    const w = x1 - x0;

    ctx.save();

    // Draw zooming cone that connects previous level to this one
    ctx.fillStyle = colors.zoomCone;
    if (level > 0 && t > dipTime) {
      const ytop = scaleY(0) + treeCellH + 1;
      ctx.beginPath();
      ctx.moveTo(scaleX0(0), ytop);
      ctx.lineTo(x0, y);
      ctx.lineTo(x1, y);
      ctx.lineTo(scaleX1(0), ytop);
      ctx.fill();
    }

    // Translate to the topleft corner of box
    ctx.translate(x0, y);

    // Make opaque
    ctx.fillStyle = "#fff";
    ctx.fillRect(0, 0, w, treeCellH);

    // Draw inner cells
    if (t === 1) {
      ctx.save();
      for (let cell = 0; cell < numCells; cell++) {
        this.drawTreeCell(ctx, level, cell);
        ctx.translate(treeCellW, 0);
      }
      ctx.restore();
    }

    // Draw outer border
    ctx.strokeRect(0, 0, w, treeCellH);

    // Tick font
    ctx.font = "10px sans-serif";
    ctx.textBaseline = "bottom";
    ctx.textAlign = "center";

    // Draw the major _unix epoch_ and _now_ ticks
    if (level === 0) {
      const drawTick = (t, color, title) => {
        const x = treeTimeX[level](t);
        if (x < 0 || x > treeColX(numCells)) return;
        ctx.fillStyle = ctx.strokeStyle = color;
        ctx.beginPath();
        ctx.moveTo(x, treeRowY(-0.7));
        ctx.lineTo(x, treeRowY(1));
        ctx.stroke();
        ctx.fillText(title, x, treeRowY(-2));
      };
      drawTick(0, colors.unixEpoch, "unix epoch");
      drawTick(+new Date() * 1e6, colors.now, "now");
    }

    // Draw the date ticks
    if (t === 1) {
      ctx.strokeStyle = ctx.fillStyle = colors.dateTick;
      const drawTick = (t, title) => {
        const x = treeTimeX[level](t);
        ctx.beginPath();
        ctx.moveTo(x, treeRowY(0));
        ctx.lineTo(x, treeRowY(-0.5));
        ctx.stroke();
        ctx.fillText(title, x, treeRowY(-0.75));
      };
      const count = 8;
      const ticks = treeTimeX[level].ticks(count);
      const tickFormat = treeTimeX[level].tickFormat(count);
      for (let i = 0; i < ticks.length; i++) {
        const tickTime = ticks[i];
        // nanosecond ticks sometimes duplicate
        if (ticks[i] !== ticks[i - 1]) {
          drawTick(tickTime, tickFormat(tickTime));
        }
      }
    }

    // Draw the node length
    if (t === 1) {
      ctx.textAlign = "left";
      ctx.textBaseline = "middle";
      ctx.fillText(nodeLengthLabels[level], x1 + 28, treeRowY(0.5));
    }

    ctx.restore();
  };
  getTreeHeight = () => {
    const { path, levelOffset, treeCellH } = this.state;
    return (path.length - 1) * levelOffset * treeCellH + treeCellH;
  };
  drawScrubGuide = ctx => {
    const { scrubbing, path, pathAnim } = this.state;
    if (!scrubbing) return;
    ctx.save();
    ctx.strokeStyle = colors.scrub;
    ctx.translate(-30, 0);
    const barH = 32;
    const treeH = this.getTreeHeight();
    ctx.beginPath();
    ctx.moveTo(0, 0);
    ctx.lineTo(0, treeH);
    ctx.lineWidth = 1;
    ctx.stroke();
    const y = d3scale
      .scaleLinear()
      .domain([1, path.length])
      .range([0, treeH - barH])(pathAnim);
    ctx.beginPath();
    ctx.moveTo(0, y);
    ctx.lineTo(0, y + barH);
    ctx.lineWidth = 4;
    ctx.stroke();
    ctx.restore();
  };
  drawTree = ctx => {
    ctx.save();
    const { treeX, treeY, treeCellH, levelOffset, path } = this.state;
    ctx.translate(treeX, treeY);
    // this.drawScrubGuide(ctx);
    for (let level = 0; level < path.length; level++) {
      this.drawTreeNode(ctx, level);
      ctx.translate(0, treeCellH * levelOffset);
    }
    ctx.restore();
  };
  drawCalendarCell = (ctx, level, cell) => {
    const s = this.state.calCellSize;
    const len = 4;
    ctx.save();
    ctx.beginPath();
    ctx.moveTo(0, 0);
    ctx.lineTo(s, 0);
    ctx.stroke();
    ctx.globalAlpha *= 0.5;
    ctx.beginPath();
    ctx.moveTo(0, 0);
    ctx.lineTo(0, len);
    ctx.stroke();
    ctx.restore();
  };
  drawCalendarNode = (ctx, level) => {
    const n = this.state.numSquareCells;
    const s = this.state.calCellSize;
    let cell = 0;
    ctx.save();
    for (let cy = 0; cy < n; cy++) {
      ctx.save();
      for (let cx = 0; cx < n; cx++) {
        this.drawCalendarCell(ctx, level, cell++);
        ctx.translate(s, 0);
      }
      ctx.restore();
      ctx.translate(0, s);
    }
    ctx.restore();
  };
  drawCalendarNodeTicks = (ctx, level) => {
    ctx.save();
    ctx.lineWidth *= 2;
    ctx.font = "10px sans-serif";
    ctx.textBaseline = "top";
    ctx.textAlign = "left";
    const pad = { left: 5, top: 4 };
    const { calTimeK, calKX, calKRow, calRowY } = this.ds;
    if (level === 0) {
      const drawTick = (t, color, title) => {
        const k = calTimeK[level](t);
        const x = calKX(k);
        const row = calKRow(k);
        ctx.fillStyle = ctx.strokeStyle = color;
        ctx.beginPath();
        ctx.moveTo(x, calRowY(row));
        ctx.lineTo(x, calRowY(row + 1));
        ctx.stroke();
        let y = calRowY(row);
        for (let text of title.split("\n")) {
          ctx.fillText(text, x + pad.left, y + pad.top);
          y += 12;
        }
      };
      drawTick(0, colors.unixEpoch, "unix\nepoch");
      drawTick(+new Date() * 1e6, colors.now, "now");
    }

    ctx.lineWidth /= 2;
    ctx.strokeStyle = ctx.fillStyle = colors.dateTick;
    const drawTick = (t, title) => {
      const k = calTimeK[level](t);
      const x = calKX(k);
      const row = calKRow(k);
      ctx.beginPath();
      ctx.moveTo(x, calRowY(row));
      ctx.lineTo(x, calRowY(row + 1));
      ctx.stroke();
      let y = calRowY(row);
      for (let text of title.split("\n")) {
        ctx.fillText(text, x + pad.left, y + pad.top);
        y += 12;
      }
    };
    const count = 32;
    const ticks = calTimeK[level].ticks(count);
    const tickFormat = calTimeK[level].tickFormat(count);
    for (let i = 0; i < ticks.length; i++) {
      const tickTime = ticks[i];
      if (ticks[i] === ticks[i - 1]) continue; // nanosecond ticks sometimes duplicate
      if (level === 0 && i === 7) continue; // we already drew this as "unix epoch"
      const text = tickFormat(tickTime)
        .split(" ")
        .join("\n");
      drawTick(tickTime, text);
    }
    ctx.restore();
  };
  drawCalendar = ctx => {
    const {
      path,
      pathAnim,
      numSquareCells,
      calCellSize,
      calX,
      calY
    } = this.state;
    const { dipTime, calW, calTimeK } = this.ds;

    const s = calCellSize;
    const n = numSquareCells;

    const t = pathAnim % 1;

    // time used to control the camera
    let camT = d3scale
      .scaleLinear()
      .domain([dipTime, 1])
      .range([0, 1])
      .clamp(true)(t);

    // time used to control the child highlight border
    let highlightT = d3scale
      .scaleLinear()
      .domain([0, dipTime / 2])
      .range([0, 1])
      .clamp(true)(t);

    let index = Math.floor(pathAnim);
    let contextLabelIndex = index - 1;

    // edge case for last path
    if (index > path.length - 1) {
      index = path.length - 1;
      contextLabelIndex = index;
      camT = 1;
      highlightT = 1;
    }

    const child = path[index];
    const level = index - 1;

    const treeColX = child % n;
    const treeRowY = Math.floor(child / n);

    const [x, y, scale] = d3scale
      .scaleLinear()
      .domain([0, 1])
      .range([[treeColX * s, treeRowY * s, 1 / n], [0, 0, 1]])(camT);

    ctx.save();
    ctx.translate(calX, calY);

    // draw date context
    ctx.save();
    ctx.translate(calW, 0);
    ctx.textBaseline = "bottom";
    ctx.textAlign = "right";
    const contextDate = calTimeK[level].contextFormat();
    if (contextDate) {
      ctx.fillStyle = "rgba(90,110,100, 0.3)";
      ctx.fillText(contextDate, 0, -10);
    }
    ctx.restore();

    // draw node time length
    ctx.save();
    ctx.translate(calW, calW);
    ctx.textBaseline = "top";
    ctx.textAlign = "right";
    ctx.fillStyle = "rgba(90,110,100, 0.5)";
    ctx.fillText(nodeLengthLabels[contextLabelIndex], -5, 10);
    ctx.restore();

    // clip window
    ctx.beginPath();
    ctx.rect(-1, -1, calW + 2, calW + 2);
    ctx.clip();

    const transformToChild = () => {
      ctx.translate(x, y);
      ctx.scale(scale, scale);
      ctx.lineWidth /= scale;
    };
    const transformToParent = () => {
      transformToChild();
      ctx.scale(n, n);
      ctx.translate(-treeColX * s, -treeRowY * s);
      ctx.lineWidth /= n;
    };

    // draw parent
    const gridColor = d3interpolate.interpolate("#fff", colors.cellWall)(0.3);
    ctx.strokeStyle = gridColor;
    ctx.save();
    transformToParent();
    this.drawCalendarNode(ctx, level);
    ctx.restore();

    // draw parent ticks
    ctx.save();
    transformToParent();
    this.drawCalendarNodeTicks(ctx, level);
    ctx.strokeRect(0, 0, calW, calW);
    ctx.restore();

    // draw child
    const childAlpha = d3interpolate.interpolate(0, 1)(Math.pow(camT, 2));
    ctx.save();
    transformToChild();
    ctx.beginPath();
    ctx.rect(-1, -1, calW + 2, calW + 2);
    ctx.clip();
    ctx.globalAlpha *= childAlpha;
    ctx.fillStyle = "#fff";
    ctx.fillRect(0, 0, calW, calW);
    this.drawCalendarNode(ctx, level + 1);
    this.drawCalendarNodeTicks(ctx, level + 1);
    ctx.restore();

    // outline child
    ctx.save();
    transformToChild();
    ctx.strokeStyle = d3interpolate.interpolate(
      "rgba(0,0,0,0)",
      colors.cellWall
    )(highlightT);
    ctx.strokeRect(0, 0, calW, calW);
    ctx.restore();

    // outline window
    ctx.strokeStyle = colors.cellWall;
    ctx.strokeRect(0, 0, calW, calW);

    ctx.restore();
  };
  draw = canvas => {
    if (!canvas) return;
    const ctx = canvas.getContext("2d");
    const { pixelRatio } = this.ds;
    const { width, height } = this.state;
    ctx.save();
    ctx.scale(pixelRatio, pixelRatio);
    ctx.clearRect(0, 0, width, height);
    this.drawTree(ctx);
    // this.drawCalendar(ctx);
    ctx.restore();
  };
  getMousePos = e => {
    if (!e) return this.lastMouse || { x: 0, y: 0 };
    const rect = this.canvas.getBoundingClientRect();
    const x = e.clientX - rect.left;
    const y = e.clientY - rect.top;
    return (this.lastMouse = { x, y });
  };
  isLevelVisible = level => {
    return level < Math.floor(this.state.pathAnim);
  };
  isCellExpanded = (level, cell) => {
    return this.state.path[level + 1] === cell;
  };
  mouseToPath = (x, y) => {
    const { treeX, treeY, treeCellW, treeCellH, levelOffset } = this.state;

    const cell = Math.floor((x - treeX) / treeCellW);
    const gridY = Math.floor((y - treeY) / treeCellH);
    const level = Math.floor(gridY / levelOffset);
    if (
      cell >= 0 &&
      cell < 64 &&
      gridY % levelOffset === 0 &&
      this.isLevelVisible(level)
    ) {
      return { level, cell };
    }

    // TODO: compute "derived" resolution cells
  };
  scrubAnim = (x, y) => {
    const { treeY, path } = this.state;
    const t = d3scale
      .scaleLinear()
      .domain([treeY, treeY + this.getTreeHeight()])
      .range([1, path.length])
      .clamp(true)(y);
    this.setState({ pathAnim: t });
  };
  onMouseDown = (e, { isDrag }) => {
    const { x, y } = this.getMousePos(e);
    const cellHighlight = this.mouseToPath(x, y);
    if (cellHighlight) {
      const { level, cell } = cellHighlight;
      if (this.isLevelVisible(level + 1) && !this.isCellExpanded(level, cell)) {
        const path = this.state.path.slice(0, level + 1);
        path.push(cell);
        this.setState({ path, pathAnim: level + 2 });
      }
      this.mousedownLevel = level;
      if (!isDrag) {
        this.mousedownCell = cell;
        this.shouldCollapseOnMouseUp = this.isCellExpanded(level, cell);
      } else if (cell !== this.mousedownCell) {
        this.shouldCollapseOnMouseUp = false;
      }
    }
  };
  cancelTransitions = () => {
    d3transition.interrupt("collapse-cell");
    d3transition.interrupt("expand-cell");
    d3transition.interrupt("collapse-all");
    d3transition.interrupt("expand-all");
  };
  onMouseUp = e => {
    this.mousedownLevel = null;
    const { cellHighlight, pathAnim } = this.state;
    if (cellHighlight) {
      const { level, cell } = cellHighlight;
      const parentPath = this.state.path.slice(0, level + 1);
      if (this.isLevelVisible(level + 1)) {
        const interp = d3interpolate.interpolate(pathAnim, level + 1);
        if (this.shouldCollapseOnMouseUp) {
          this.cancelTransitions();
          d3transition
            .transition("collapse-cell")
            .duration(500)
            .tween("pathAnim", () => t =>
              this.setState({ pathAnim: interp(t) })
            )
            .on("end", () => this.setState({ path: parentPath }));
        }
      } else {
        parentPath.push(cell);
        this.setState({ path: parentPath });
        this.cancelTransitions();
        d3transition
          .transition("expand-cell")
          .duration(500)
          .tween("pathAnim", () => t =>
            this.setState({ pathAnim: level + 1 + t })
          );
      }
    }
  };
  onMouseMove = e => {
    const { x, y } = this.getMousePos(e);
    if (this.state.scrubbing) {
      this.scrubAnim(x, y);
    } else {
      const curr = this.mouseToPath(x, y);
      const prev = this.state.cellHighlight;
      this.setState({
        cursor: curr ? "pointer" : "default"
      });
      if (JSON.stringify(curr) !== JSON.stringify(prev)) {
        this.setState({ cellHighlight: curr });
      }
      if (curr && this.mousedownLevel === curr.level) {
        this.onMouseDown(e, { isDrag: true });
      }
    }
  };
  onKeyDown = e => {
    if (e.key === "Shift") {
      this.setState({ scrubbing: true });
      this.setState({ cursor: "grabbing" });
      this.onMouseMove();
    }
  };
  onKeyUp = e => {
    if (e.key === "Shift") {
      this.setState({ scrubbing: false });
      this.setState({ cursor: "default" });
    } else if (e.key === "Enter") {
      this.cancelTransitions();
      const { pathAnim, path } = this.state;
      const durationPerLevel = 125;
      if (pathAnim > 1) {
        const interp = d3interpolate.interpolate(pathAnim, 1);
        const dist = pathAnim - 1;
        const duration = dist * durationPerLevel;
        d3transition
          .transition("collapse-all")
          .ease(d3ease.easeLinear)
          .duration(duration)
          .tween("pathAnim", () => t => this.setState({ pathAnim: interp(t) }));
      } else {
        const interp = d3interpolate.interpolate(pathAnim, path.length);
        const dist = path.length - pathAnim;
        const duration = dist * durationPerLevel;
        d3transition
          .transition("expand-all")
          .ease(d3ease.easeLinear)
          .duration(duration)
          .tween("pathAnim", () => t => this.setState({ pathAnim: interp(t) }));
      }
    }
  };
  getCssCursor = cursor => {
    if (cursor === "grabbing") {
      cursor = "-webkit-grabbing";
    }
    return cursor;
  };
  render() {
    const { width, height, cursor } = this.state;
    const { pixelRatio } = this.ds;
    return (
      <canvas
        ref={node => {
          this.canvas = node;
          this.draw(node);
        }}
        width={width * pixelRatio}
        height={height * pixelRatio}
        style={{
          width: `${width}px`,
          height: `${height}px`,
          userSelect: "none",
          cursor: this.getCssCursor(cursor)
        }}
        onMouseMove={this.onMouseMove}
        onMouseDown={e => this.onMouseDown(e, { isDrag: false })}
        onMouseUp={this.onMouseUp}
        onDragStart={() => false}
      />
    );
  }
}

export default Viz;
